### PR TITLE
Refactor figure generation

### PR DESCRIPTION
**Pull request by kaitj**
_(rebased to updated dependencies on July 25, 2024)_

This PR follows up from #10 - I've included some of the discussion around this in the dropdown below:

<details><summary>Figure discussion</summary>

>Member
@[clane9](https://github.com/clane9) clane9 [2 days ago](https://github.com/childmindresearch/niftyone/pull/10#discussion_r1608470669)
> Leaving a note from our offline conversation. Currently two figures plus qc metrics are being generated here. We discussed decoupling this a bit more. Specifically, we discussed:
> 
> Separating the concerns of figure generation and QC metric extraction.
> Uncoupling the figures so that there is one generator per figure
> Assigning the responsibility of filtering the records to the generator itself. (Currently that responsibility lies outside in the calling pipeline.)
> One possible abstraction we discussed was:
> 
> ```python
> class Generator:
>     def __init__(self, filters: dict[str, Any]):
>         self.filters = filters
> 
>     def __call__(self, table: BIDSTable, out_dir: Path, overwrite: bool = False):
>         """Filter table and generate figures for each matching record."""
>         table = table.filter_multi(**self.filters)
>         for _, record in table.iterrows():
>             self.generate(record, out_dir, overwrite)
> 
>     def generate(self, record: pd.Series, out_dir: Path, overwrite: bool = False):
>         """Generate figure for record."""
>         raise NotImplementedError
> 
> 
> class Pipeline:
>     def __init__(self, generators: list[Generator]):
>         self.generators = generators
> 
>     def run(self, table: BIDSTable, out_dir: Path, overwrite: bool = False):
>         for gen in self.generators:
>             gen(table, out_dir, overwrite)
>  ```
> For configuring pipelines, we could define the list of figure generators (and their kwargs) in a yaml config. Then at runtime, construct the list of generators and run the pipeline.
> 
>  ```yaml
> figures:
>   three_view:
>     filters:
>       suffix: T1w
>   slice_video:
>     filters:
>       suffix: T1w
>   three_view_video:
>     filters:
>       suffix: bold
>   carpet_plot:
>     filters:
>       suffix: bold
>   mean_std:
>     filters:
>       suffix: bold
>  ```
> The generator lookup could be handled with a global registry, with generators registered via a decorator when the class is defined.
> 
> I'm not fully convinced this is the best abstraction, so I'm open to alternatives. In particular, the config looks quite repetitive. This is not necessarily the worst thing, since decoupling is worth some repetition. But it might be a sign we are missing a better idea.
> 
> Maybe something along the lines of letting groups of figures be associated with each set of BIDS filters, rather than a 1-1 mapping of filters to figures.
> 
> @[nx10](https://github.com/nx10) thoughts?
> 
> Member
> Author
> @[kaitj](https://github.com/kaitj) kaitj [2 days ago](https://github.com/childmindresearch/niftyone/pull/10#discussion_r1608639650) • 
> Hmm, another (not fully flushed out) thought if going the yaml route:
> 
> ```yaml
> key1:
>   filters:
>       desc: [True, brain, False]
>       suffix: [T1w, T2w, T1map]
>   views: [three_view, slice_video]
> 
> key2:
>   filters:
>       suffix: bold
>   views: [three_view_video, carpet_plot, mean_std]
> ```
> The entities provided in filters are zipped and passed along together. If an entity True, it treats that entity as a wildcard (e.g), if False, exclude rows that contain an entity for that filter. Otherwise, the filter has to match exactly. If an entity isn't provided, we can also treat it similar to False (subject entity aside). So in the above case, it would generate the views for sub-<subject>_desc-*_T1w, sub-<subject>_desc-brain_T2w, and sub-<subject>_T1map.
> 
> The views will indicate the figure(s) that should be generated for those matching filters. Figure generation time aside, the end-user can always ignore those in NiftyOne itself. The primary keys (e.g. key1, key2) are kind of arbitrary and really are only there to group filters with figures to generate. I don't think that the repetitiveness is necessarily a bad thing, but also don't want it to get too unruly.
> 
> As an aside, we're going to need some good documentation so that a user can find the different views that can be generated (irrespective of the implementation method) 😅 .
> 
> And also open to other ideas of course.
> 
> Member
> @[nx10](https://github.com/nx10) nx10 [2 days ago](https://github.com/childmindresearch/niftyone/pull/10#discussion_r1608679370)
> Yeah either use groups of matrices like @kaitj suggests (would be my preference), or consider "inverting" that tree:
> 
> ```yaml
> figures:
>   t1w: [three_view, slice_video]
>   bold: [three_view_video, carpet_plot, mean_std]
> ```
> Member
> @[nx10](https://github.com/nx10) nx10 [2 days ago](https://github.com/childmindresearch/niftyone/pull/10#discussion_r1608702887)
> Overall I would prefer a [Pandas query string](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html) or [Polars SQL query](https://docs.pola.rs/py-polars/html/reference/api/polars.SQLContext.execute.html) (for SQL I would just let the user set SELECT * FROM df WHERE {userinput}).
> 
> It's always tempting to reimplement a small query DSL because one thinks they don't need all of that complexity, but then slowly realize that the complexity is needed.
> 
> Member
> @[nx10](https://github.com/nx10) nx10 [2 days ago](https://github.com/childmindresearch/niftyone/pull/10#discussion_r1608712896) • 
> With @kaitj s matrix groups this would look like:
> 
> ```yaml
> my_t1w_images:
>   query: "( desc in (None, 'brain') ) and ( suffix in ('T1w', 'T2w', 'T1map') )"
>   views: [three_view, slice_video]
> 
> my_bold_images:
>   query: "suffix == 'bold'"
>   views: [three_view_video, carpet_plot, mean_std]
> ```
> Group name could also become a tag, then the same logic can be used to generate user defined tags (potentially without views):
> 
> ```yaml
> really_bad_images:
>   query: "whatever_qc_metric < 0"
> ```
> Member
> @[clane9](https://github.com/clane9) clane9 [2 days ago]
> Thanks @kaitj, @nx10 for the good suggestions. Let's go with your guys' preferred direction. I also think having a query option is a good idea. Having a simple way to do basic filters might also be a good idea, but it's less clear to me what the api for this should be. So I think let's just implement the query option for now and decide later on if there are common verbose query patterns we would like to streamline.
> 
> The suggestion for QC metric queries is also interesting. On the one hand, I can imagine a use case for generating figures conditional on some QC value. E.g. certain expensive to generate detailed figures might only be useful for bad or borderline data. On the other, it could be a bit of an anti-pattern since filtering by QC values can happen naturally within the fiftyone app. Also, it would require joining the QC metric table(s) with the b2t table prior to figure generation. (Not a big deal but something to note.) @kaitj what do you think about QC based queries?
> 
> Member
> Author
> @[kaitj](https://github.com/kaitj) kaitj [2 days ago](https://github.com/childmindresearch/niftyone/pull/10#discussion_r1608742571)
> Yeah, I like the query option too. I think it'll be familiar too (hopefully) for what I imagine would be the majority of users.
> 
> Regarding the QC metric queries, I think we are on a similar page. I can see a use case for it too. In fact, it is likely what we would want to do for identifying problematic images as you noted, but I also think it is probably better to leave that kind of filteirng for within the app or potentially for power(?) users (e.g. interactive plot sort of thing).

</details>

First crack at refactoring the figure creations to using a generator for creating the various figures and a "Runner" that actually calls these functions within the analysis workflows. Should be able to extend it to custom figure generation code (see test cases). At the moment it is only doing the figure generation, but I think the metric updates will also get thrown in there. Not sure about the group-level yet, but may also be able to reuse it there. 

We had also chatted offline the other day about a yaml config. I ended up writing one and adding into the `resources` (more as a way to point a user to an example and to use as a default). From an api standpoint, we could directly add support for a nested dictionary config instead of loading a yaml file. For the CLI, I think it's easier to ingest a yaml file than potentially complex optional argument(s).

This PR strips away the QC metric functionality and I am in the process of adding that back in (will be separate PR) while respecting the config that gets passed.

### CODE REVIEW

**Comment by clane9** [comment]
Thanks @kaitj! I just merged #10. Can you rebase and then I can take a look at the changes?

File: src/niftyone/figures/generator.py
Comment (nx10) [inline src/niftyone/figures/generator.py:22]: Why `Type[T]` instead of `T`?

**Comment by kaitj** [comment]
Sounds good. I think once the kinks of this PR are resolved, would still want to hold off on merging until the metrics stuff is also ready.

File: src/niftyone/figures/generator.py
Comment (kaitj) [reply src/niftyone/figures/generator.py:22]: `Type[T]` for working with the (sub)class directly. The registration function does this, registering the generator classes. Unless I am misunderstanding?

Comment (clane9) [reply src/niftyone/figures/generator.py:22]: I do not understand why `T` and not `Type["ViewGenerator"]`? Instances of type `T` here are instances of `ViewGenerator` or its subclasses, no?

File: src/niftyone/figures/generator.py
Comment (clane9) [inline src/niftyone/figures/generator.py:38]: Here the keys of `config` are discarded. Should we keep them and return a `dict[str, "ViewGenerator"]`?

File: src/niftyone/figures/generator.py
Comment (nx10) [reply src/niftyone/figures/generator.py]: Just using ViewGenerator would be the non generic version. `Type[]` is not needed at all.

File: src/niftyone/figures/generator.py
Comment (clane9) [inline src/niftyone/figures/generator.py:44]: I might have this be a separate function `create_generator`. It could accept common args as well as arbitrary `**kwargs` since generators may have their own options. It could also handle the error handling for unregistered generators.

File: src/niftyone/figures/generator.py
Comment (clane9) [inline src/niftyone/figures/generator.py:48]: Should a missing generator be an exception not a warning?

**Review by clane9 — COMMENTED** [review]
Left a few initial comments. Will add more later on.

File: src/niftyone/figures/generator.py
Comment (kaitj) [reply src/niftyone/figures/generator.py:46]: We could raise an exception - I guess the question is do we want the workflow to "stop" if it can't create the view?

File: src/niftyone/figures/generator.py
Comment (kaitj) [reply src/niftyone/figures/generator.py:36]: Hmm, I wasn't sure what to use the keys here for since they seem arbitrary and are really only used to group the query / views. I do like the idea of keeping them around more I think about it.

File: src/niftyone/figures/generator.py
Comment (kaitj) [reply src/niftyone/figures/generator.py:42]: This took a little bit of hacking at, but what do you think of being able to pass the kwargs in via the config this way:

```yaml
anat:
  query: datatype == 'anat'
  views: ['three_view', 'slice_video(axis=2,vmin=0,vmax=1)']
```

If its provided, we'll parse it internally to split kwargs from the registered view. This would be an internal call within `create_generator`.

File: src/niftyone/figures/generator.py
Comment (kaitj) [reply src/niftyone/figures/generator.py:46]: I've made this to just raise a custom `GeneratorError` for now within the `create_generator`call.

File: src/niftyone/figures/generator.py
Comment (kaitj) [reply src/niftyone/figures/generator.py]: Set this back to `type[T]`. 

re: @nx10's comment, had an offline discussion, largely around handling `mypy` throwing an error. Something about mypy not understanding inheritance where `type[B] != type[A]` even though `B` is a subclass of `A`. 

re: the `type["ViewGenerator"]` comment, while `mypy` is happy with this, Python isn't fully type-safe. It should probably error out here since subclasses would be instantiated as "ViewGenerator" and in theory, we shouldn't be able to call methods of the subclass. 

Hopefully that made sense, still trying to wrap my head around the second point.

File: src/niftyone/figures/generator.py
Comment (clane9) [reply src/niftyone/figures/generator.py]: I kind of prefer to just raise builtin exceptions wherever possible. Less code and one less exception class to think about handling. I would just raise `KeyError`.

File: src/niftyone/figures/generator.py
Comment (clane9) [reply src/niftyone/figures/generator.py]: Hm I am not sure about letting the entries in `views` be this complex. Now wouldn't building the generators require either some more involved string parsing or `eval`? What about

```
anat:
  query: datatype == 'anat'
  views:
    - name: three_view
    - name: slice_video
      axis: 2
      vmin: 0
      vmax: 1
```

Comment (kaitj) [reply src/niftyone/figures/generator.py]: Ahh yeah, I do like this more (think I've been staring at this code too much the last few days 😅). You're right that it would require more involved string parsing otherwise.

I'll tighten this up a little bit later today with this in mind.

File: src/niftyone/figures/multi_view.py
Comment (clane9) [inline src/niftyone/figures/multi_view.py:14]: The repeated use of `generate_common` feels like it could be simplified.

File: src/niftyone/figures/generator.py
Comment (nx10) [reply src/niftyone/figures/generator.py]: What happens if the view has a Parameter `name`?

Comment (kaitj) [reply src/niftyone/figures/generator.py]: Would `name` be associated with anything other than the view? I could see it maybe being a figure title, but I think it's more likely that any figure related kwargs use those associated kwargs (e.g. `title`, `xlabel`, `ylabel`, etc.).

File: src/niftyone/figures/multi_view.py
Comment (kaitj) [reply src/niftyone/figures/multi_view.py:14]: Hmm, if there is a way to use `partial` with it, that could work. Will need to look into this.

File: src/niftyone/figures/generator.py
Comment (kaitj) [reply src/niftyone/figures/generator.py]: Done, happy with using the built-in exceptions. Not sure if we would have to handle any other errors at the moment. I think if there are a lot more that are related to the generator, my preference would be to go back to a custom exception.

File: src/niftyone/figures/multi_view.py
Comment (clane9) [reply src/niftyone/figures/multi_view.py:14]: I can mess around with it a bit and work up an alternative in a separate branch for you to weigh in on.

**Comment by kaitj** [comment]
Made one small update here to pass the entities to the generators as a dict (e.g. `entities={"desc": "...", "ext": ".png"}`), which should allow it to be more flexible if we need it to be in the future. Previously, it only accepted "desc" and "ext" as arguments.

File: src/niftyone/figures/generator.py
Comment (kaitj) [reply src/niftyone/figures/generator.py]: I think I am inclined to leave this as is here unless we have a use case for the keys.

File: src/niftyone/figures/multi_view.py
Comment (kaitj) [reply src/niftyone/figures/multi_view.py]: Okay, from the repeated use of `generate_common` and `generate` front, this should be fixed in 
6de7f62. 

I haven't touched the registry - I know there was some discussion around refactoring that as well. Do we want to try to add that in here or push that down the road a little bit to try to get some additional figures generated for derivative data? While I don't love the use of decorators here anymore (given the unexpected behaviour I had encountered trying to fix `generate`), I don't think it is doing anything breaking.

**Comment by kaitj** [comment]
I am also going to mark this as ready for review. I think the core functionality is there, just a question about the registry refactoring (see response to comment above).

File: pyproject.toml
Comment (nx10) [inline pyproject.toml:8]: Personal email?

File: src/niftyone/analysis_levels/participant.py
Comment (nx10) [inline src/niftyone/analysis_levels/participant.py:11]: I think they removed this one front Setuptools (or deprecated it ?) importlib.metadata / importlib.resource I think are the new apis

Comment (kaitj) [reply src/niftyone/analysis_levels/participant.py]: This is fixed in 7aaa470. Had to add a `type: ignore` here for mypy with different expected types than what resources returned.

File: pyproject.toml
Comment (kaitj) [reply pyproject.toml:8]: I left this as is from the original and never updated it. We could update this - I'll leave it up to @clane9.

File: src/niftyone/figures/generator.py
Comment (kaitj) [reply src/niftyone/figures/generator.py]: Coming back to this and the last point here. I've [implemented a more generic overlay generator](https://github.com/kaitj/niftyone/pull/2/commits/969e85c7534ac190ee4496bfa6bb860bd992e319), which assumes that the end user is generating figures with just up-to-date implementation (e.g. via config). This generator could be used to generate overlay with masks (assuming similar entities) or be used to generate overlays to check registration (e.g. fmri/dmri-to-T1w which would be using different entities).

The problem with the current config and the implementation above is that there is a high likelihood of a generated image being overwritten since the generator entities are static (e.g. they would all have the same description and extension). 

We don't have to fix it in this PR, but wanted to open the discussion (and can move it offline or to a new issue) about how to best handle this with the config.

File: pyproject.toml
Comment (clane9) [reply pyproject.toml:8]: This is just the email I use for github. I don't see a strong reason to change. But I do think @kaitj you should add yourself as another author.

Comment (clane9) [reply pyproject.toml:8]: Also in #16 I believe there was perhaps a ruff update that applied these nonsignificant formatting diffs. Can we rebase/ fix the ruff version so we don't undo these diffs?

Comment (kaitj) [reply pyproject.toml:8]: I'll take a look - you're right, there was likely an update in the formatting somewhere.

Comment (kaitj) [reply pyproject.toml]: Okay, just made a new commit that adds myself as another author. Re: formatting - I fixed it in the same commit, but its not a ruff issue. It's my vscode "formatOnSave" feature that is changing the formatting.

Comment (nx10) [reply pyproject.toml]: I see, I usually try to avoid that to get less spam from bots crawling GitHub. But 🤷

Comment (clane9) [reply pyproject.toml]: That's why I use this email. It's a trash public email that gets a lot of spam. What is your suggestion @nx10?

Comment (nx10) [reply pyproject.toml]: Whatever you are comfortable with, I was planning on changing all of mine to dair@childmind.org soon. I don't think it matters outside of spam protection.

**Comment by kaitj** [comment]
Adding this as a note as it came up during discussion with @clane9 yesterday. 

We may want to protect the standard bids entities as they may have meaning associated to each file. Instead, we probably want to use an "extra entity" (e.g. `figure`). I took a quick look through the current bids standard and extension proposals, but couldn't find anything that might guide us in a direction for this. For now, I am just using `figure` as the default entity (though some of these are videos 🤷‍♂️) - open to suggestions here.

Additionally, we may want to make this entity customizable instead of static to avoid different outputs with otherwise identical entities being overwritten (e.g. currently if we were to say generate 2 carpet plots with the same entities, only 1 carpet plot would exist; this example is probably better digested with overlays). For now, I've added these into the view_kwargs with figure being protected for the image naming. Currently, the only downside I see to this is that "figure" has to be included in the `view_fn` without actually being used for anything. From our brief discussions, this may ultimately be another refactor to handle this differently, but I don't want to prevent this from being a blocking issue with the current implementation.

Both of these changes can be seen in 7e11522.

**Review by nx10 — APPROVED** [review]
🚀

**Pull request merged**
